sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -606,301 +576,439 @@ abstract class FileParquetPartitionReaderBase(
+     * If there was an error then the error field is set. If there were no blocks the buffer
+     * is returned as null.  If there were no columns but rows (count() operation) then the
+     * buffer is null and the size is the number of rows.
+     *
+     * Note that the TaskContext is not set in these threads and should not be used.
+     */
+    override def call(): HostMemoryBuffersWithMetaData = {
+      val hostBuffers = new ArrayBuffer[(HostMemoryBuffer, Long)]
+      try {
+        val fileBlockMeta = filterHandler.filterBlocks(file, conf, filters, readDataSchema)
+        if (fileBlockMeta.blocks.length == 0) {
+          // no blocks so return null buffer and size 0
+          return HostMemoryBuffersWithMetaData(fileBlockMeta.isCorrectedRebaseMode,
+            fileBlockMeta.schema, fileBlockMeta.partValues, Array((null, 0)), file.filePath,
+            file.start, file.length, None)
+        }
+        blockChunkIter = fileBlockMeta.blocks.iterator.buffered
+        if (isDone) {
+          // got close before finishing
+          HostMemoryBuffersWithMetaData(
+            fileBlockMeta.isCorrectedRebaseMode,
+            fileBlockMeta.schema, fileBlockMeta.partValues, Array((null, 0)),
+            file.filePath, file.start, file.length, None)
+        } else {
+          if (readDataSchema.isEmpty) {
+            val numRows = fileBlockMeta.blocks.map(_.getRowCount).sum.toInt
+            // overload size to be number of rows with null buffer
+            HostMemoryBuffersWithMetaData(fileBlockMeta.isCorrectedRebaseMode,
+              fileBlockMeta.schema, fileBlockMeta.partValues, Array((null, numRows)),
+              file.filePath, file.start, file.length, None)
+
+          } else {
+            val filePath = new Path(new URI(file.filePath))
+            while (blockChunkIter.hasNext) {
+              val blocksToRead = populateCurrentBlockChunk(blockChunkIter,
+                maxReadBatchSizeRows, maxReadBatchSizeBytes)
+              val blockTotalSize = blocksToRead.map(_.getTotalByteSize).sum
+              hostBuffers += readPartFile(blocksToRead, fileBlockMeta.schema, filePath)
+            }
+            if (isDone) {
+              // got close before finishing
+              hostBuffers.foreach(_._1.close())
+              HostMemoryBuffersWithMetaData(fileBlockMeta.isCorrectedRebaseMode,
+                fileBlockMeta.schema, fileBlockMeta.partValues, Array((null, 0)),
+                file.filePath, file.start, file.length, None)
+            } else {
+              HostMemoryBuffersWithMetaData(fileBlockMeta.isCorrectedRebaseMode,
+                fileBlockMeta.schema, fileBlockMeta.partValues, hostBuffers.toArray,
+                file.filePath, file.start, file.length, None)
+            }
+          }
+        }
+      } catch {
+        case e: Exception =>
+          if (!isDone) {
+            logError(s"Exception in Parquet file read thread, ${e.getMessage}", e)
```

Comment:
Do we want to log here and also return the error?  Seems like we are doing an indirect version of the log-then-throw anti-pattern here.  Also curious why we don't just let the Future throw when the submitter tries to get the value rather than wrap it explicitly here.

Reply:
ah yes thanks for pointing this out, this originated in a much different version of the code. We should be able to just let the future throw now.